### PR TITLE
Implement FIRE_IMMUNE flag and grant it to fire entry suit and activated RM13 combat armor

### PR DESCRIPTION
**Pull request by zachary-kaelan**
#### Summary
Content "Fire entry suit grants total fire and heat immunity"

#### Purpose of change

An entry suit is the kind of thing you might wear when you expect to be completely engulfed in flames, or taking lava samples from a volcano, or getting splashed with molten metal, or defusing an incendiary grenade. There wasn't any indication that it was capable of any of those things. 

I also noticed in the code that there were a number of manual checks to see if a character was wearing RM13 combat armor, to determine fire immunity. 

#### Describe the solution

I added a new flag, FIRE_IMMUNE, and replaced all those checks with checks against that flag. I gave it to those two pieces of armor. Also, for some reason the entry suit was half Nomex and half Kevlar, so I axed the Kevlar. And somehow it only weighed 2.9 kg so I upped that to 24.9 kg based on [this](https://www.safedesign.com/products/newtex-x50-sd-fire-entry-suit).

#### Describe alternatives you've considered

Based on what I saw in the code, there are likely some redundancies and unused functions present in the checks for immunity, both pre-existing and new. I may do some refactoring in a future PR.

#### Testing

WIP.

### CODE REVIEW

**Review by kevingranade — CHANGES_REQUESTED** [review]
Looks like someone didn't know the words "bunker suit" and called the fireman gear the wrong thing.
Instead of changing fireman gear to be an entry suit, add an entry suit with appropriately nonexistent spawns.

**Comment by Terrorforge** [comment]
If the entry suit needs to be its own thing, perhaps separate out that into its own PR and let this just be the implementation of FIRE_IMMUNE. That flag would be useful in and of itself, as MoM and XE currently have fire immunity effects that have to rely on slightly hacky workarounds that individually add heat resistance, blister immunity, heat damage immunity, etc.

**Comment by Venera3** [comment]
Flat, indefinite, free fire immunity is not the right way to represent an entry suit. They buy you time, sure, but that's not something we can represent ingame.

The RM being an overblown anti-everything is a separate problem, but not one that needs expanding.

**Comment by harakka** [comment]
Despite fire immunity you will still take damage from the environment being hot, right? Your clothes will just not catch on fire.

**Comment by zachary-kaelan** [comment]
They protect from fire through reflection of radiant heat, and they protect from ambient heat through heavy insulation. But if exposed to those high temperatures for too long, especially ambient heat which doesn't get reflected, the materials start to break down. You're going to be sweating in there regardless and getting dehydrated, and you need a mask to protect from smoke. It also weighs 40 pounds (in real life) and is some of the most encumbering clothing in the game at 50.

It should definitely be not sturdy and not particularly protective, and probably shouldn't be craftable or repairable, as their construction requires applying aluminum to the surface on a molecular level using vacuum deposition.

I'll look into how to make the fire immunity temporary.

**Comment by kevingranade** [comment]
See https://www.rng.asia/post/5-things-to-know-about-aluminized-fire-proximity-suit
"Generally for the [Fire Proximity suit](https://www.rng.asia/fire-proximity-suit), complete fire engulfment time ranges between 30-45 seconds before 2nd degree burns occur,"
Other sources might have more useful information,  but reflecting away 90-95% of the incoming radiant heat just buys you time,  it doesn't make you immune.
Also note the information on heat exhaustion and ventilation and weight. Practically speaking these things will be worthless outside some very specialized use cases.

**Comment by zachary-kaelan** [comment]
There are [three types of suits](https://www.firerescue1.com/fire-products/personal-protective-equipment-ppe/articles/fire-service-proximity-suits-what-firefighters-need-to-know-8hocPfz9S5NdMdGS/) (potentially outdated numbers):

* Approach suit provides a maximum ambient heat protection of about 200 degrees F (93 degrees C).
* Proximity suit provides a maximum ambient heat protection of about 500 degrees F (260 degrees C).
* Entry suit provides a maximum ambient heat protection of about 2000 degrees F (1093 degrees C) for a short duration and prolonged radiant heat up to 1500 degrees F (816 degrees C).

But based on what I see on both that website and [this one](https://www.newtex.com/nxp/compare-quote#category/proximity-suits/), everything aluminized provides the same 3000 F (1650 C) radiant protection, with the only variation being the ambient temperature, and the duration of fire entry.

I can add a counter to anything aluminized that ticks down based on level of radiant heat using `get_heat_radiation`, before becoming compromised. There probably aren't many instances where the player would need or want to be in a fire for multiple minutes, but shorter durations would be an asset against gasoline zombies and make flamethrowers and other incendiaries more viable for clearing areas out (especially if arming NPCs). The main use case though is not being scalded to death while in mi-go towers.

**Comment by kevingranade** [comment]
Yea you really only get one chance at reducing the radiant heat. You reflect that ~90% away,  and once it's absorbed it's your problem.

Past that you're just stacking more or more effective insulating layers. The suit heats up from the outside in,  and eventually you cook. On top of that the suit defeats perspiration effects, so you're going to hit heat exhaustion even if you aren't in a fire.

Do keep in mind, these temperature ranges are just sales copy. The reality is how effective the insulation is at slowing down the heating process, so what you'd be looking at is taking incoming heat from both hot air and radiant sources and dropping it by some factor before applying it to the player.

**Comment by Terrorforge** [comment]
> Despite fire immunity you will still take damage from the environment being hot, right? Your clothes will just not catch on fire.

I thought this was the case with our current implementation, but I wanted to check. So for everyone's edification, here's how it currently works when you're standing in a raging inferno (courtesy of MoM's "Hellfire" spell, ambient temperature 400°C):

If you're butt-ass naked, stepping into the flames cripples you instantly and kills you in a few turns. If you just stand next to them, you take damage relatively quickly, dying in a matter of minutes. This is mostly from Blisters, an effect applied based on radiant heat, with a small contribution from your body temperature hitting Scorching!

If you're wearing an entry suit you can step into the flames with no ill effect, but it does nothing against the ambient heat. Heat resistance does protect against Blisters, but it's pretty all-or-nothing and when it's this hot I guess the entry suit just can't keep up.

But if you increase your heat resistance just a little further, such as by putting on a set of the "flame-resistant" undergarments in addition to the entry suit, you pass the threshold where you're immune to Blisters and only take damage very slowly from your "Scorching!" body temp. You'd be able to survive for hours, except I don't think it's possible to achieve this level of heat resistance on your mouth.

Oh, and your clothes only "catch on fire" (they don't actually; see comments on #74989) if you actually step on a tile with fire. This is not the case in real life; a raging fire can absolutely be hot enough to ignite flammable materials that aren't directly touching visible flames.

So the current implementation is very generous. You have survivability comparable to someone in an entry suit [or at least turnout gear](https://www.quora.com/How-hot-can-firefighters-get) even when naked as long as you don't actively walk into the flames, and if you really try you can get pretty close to total immunity. For a realistic entry suit to be meaningful, the effectiveness of other fire protection would have to be reduced and the general danger of fire increased. Which would be a little hazardous, because it's really easy to get too close to a fire and hurt yourself without understanding what's happening. You'd need more safeguards, like a prompt if you're about to step into an area hot enough to hurt you, and/or making hot air visible.

**Comment by kevingranade** [comment]
Ok I was actually mistaken here, this is fully intended to be an actual entry suit with all that entails, so disregard the comment about making  a new item.  However all the commentary about how an entry suit works is still pertinent.

File: src/character_body.cpp [hunk 1]
```diff
@@ -464,7 +465,8 @@ void Character::update_bodytemp()
     const bool has_bark = has_flag( json_flag_BARKY );
     const bool has_sleep = has_effect( effect_sleep );
     const bool has_sleep_state = has_sleep || in_sleep_state();
-    const bool heat_immune = has_flag( json_flag_HEAT_IMMUNE );
+    const bool fire_immune = has_flag( json_flag_FIRE_IMMUNE ) || worn_with_flag( flag_FIRE_IMMUNE );
+    const bool heat_immune = has_flag( json_flag_HEAT_IMMUNE ) || worn_with_flag( flag_HEAT_IMMUNE );
```
Comment (kevingranade) [inline src/character_body.cpp:469]: look around line 520 below:
`const units::temperature_delta h_radiation = get_heat_radiation( pos() );`
Just after that you want to adjust down h_radiation based on how good the armor is at reflecting thermal energy.
Likewise player_local_temp is acquired up on line 452, adjust that based on clothing's ability to reject heat.

**Comment by Night-Pryanik** [comment]
Closing as stale. If you wish to continue working on this, ping me to reopen.

**Pull request closed without merging**